cmd/plugin/cmd/init_test.go

```diff
@@ -0,0 +1,531 @@
+				generateGenericProvider(clusterctlv1.InfrastructureProviderType, "aws", "capa-operator-system", "", "", ""),
+				generateGenericProvider(clusterctlv1.InfrastructureProviderType, "docker", "capd-operator-system", "", "", ""),
+			},
+			wantErr: false,
+			opts: &initOptions{
+				infrastructureProviders: []string{
+					"aws:capa-operator-system",
+					"docker:capd-operator-system",
+				},
+				targetNamespace: "capi-operator-system",
+			},
+		},
+		{
+			name: "all providers",
+			wantedProviders: []genericprovider.GenericProvider{
+				generateGenericProvider(clusterctlv1.CoreProviderType, "cluster-api", "capi-system", "v1.6.0", "", ""),
+				generateGenericProvider(clusterctlv1.InfrastructureProviderType, "aws", "capa-operator-system", "", "", ""),
+				generateGenericProvider(clusterctlv1.InfrastructureProviderType, "docker", "capd-operator-system", "", "", ""),
+				generateGenericProvider(clusterctlv1.ControlPlaneProviderType, "kubeadm", "kcp-system", "", "", ""),
+				generateGenericProvider(clusterctlv1.BootstrapProviderType, "kubeadm", "bootstrap-system", "", "", ""),
+				generateGenericProvider(clusterctlv1.AddonProviderType, "helm", "caaph-system", "", "", ""),
+			},
+			wantErr: false,
+			opts: &initOptions{
+				coreProvider: "cluster-api:capi-system:v1.6.0",
+				infrastructureProviders: []string{
+					"cluster-api:capi-system:v1.6.0",
+					"aws:capa-operator-system",
+					"docker:capd-operator-system",
+				},
+				controlPlaneProviders: []string{
+					"kubeadm:kcp-system",
+				},
+				bootstrapProviders: []string{
+					"kubeadm:bootstrap-system",
+				},
+				addonProviders: []string{
+					"helm:caaph-system",
+				},
+				targetNamespace: "capi-operator-system",
+			},
+		},
+		{
+			name:            "invalid input",
+			wantedProviders: []genericprovider.GenericProvider{},
+			wantErr:         true,
+			opts: &initOptions{
+				infrastructureProviders: []string{
+					"TOO:MANY:PARTS:HERE",
+				},
+				targetNamespace: "capi-operator-system",
+			},
+		},
+		{
+			name:            "empty provider",
+			wantedProviders: []genericprovider.GenericProvider{},
+			wantErr:         true,
+			opts: &initOptions{
+				infrastructureProviders: []string{
+					"",
+				},
+				targetNamespace: "capi-operator-system",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			resources := []ctrlclient.Object{}
+
+			for _, provider := range tt.wantedProviders {
+				resources = append(resources, provider)
+			}
+
+			err := initProviders(ctx, env, tt.opts)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+
+				return
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+
+			for _, genericProvider := range tt.wantedProviders {
+				g.Eventually(func() (bool, error) {
+					provider, err := getGenericProvider(ctx, env, string(util.ClusterctlProviderType(genericProvider)), genericProvider.GetName(), genericProvider.GetNamespace())
```

Comment:
BTW, you can define a custom interface deriving from client.Object and GenericProvider to get rid of the method and reduce it to a single `client.Get`:
```go
type GeneriProviderObject {
    client.Object
    genericprovider.GenericProvider
}
…

if err := client.Get(ctx, types.NamespacedName{Name: genericProvider.GetName(), Namespace: genericProvider.GetNamespace()}, provider); err != nil {
 	return nil, err
}

```

Reply:
It says `cannot use provider (variable of type GeneriProviderObject) as client.Object value in argument to env.Get: generiProviderObject does not implement client.Object (ambiguous selector generiProviderObject.DeepCopyObject)`. Maybe we can keep it for now?